content/blog/2020-11-11-self/index.md

```diff
@@ -0,0 +1,155 @@
+
+### Type Prediction
+
+This compiler optimization builds upon Message Splitting. The motivation here seems to be mostly based on Self's operator overloading. The idea is that a message like `less-than` will often be passed to integers, floats, and strings because the language provides this behavior, but in rare cases, this message could be passed to arbitrary objects. Then, any time the compiler sees a `less-than` message passed to an object `o`, it inserts branches that check if `o` is an integer, a float, a string, or something else. Then it Message Splits the single message-pass to the ends of these four branches. This allows Message Inlining in the common case and actual message passing in the worst case.
+
+## Supporting the Programming Environment
+
+### Support for Incremental Recompilation
+
+This section of the paper addresses how the Self implementation uses dependency links between object maps and methods so that the minimal correctness-preserving adjustment can be made when either a method or object is changed by the programmer. This is like a really fine-grained version of a typical C/C++ Makefile, where dependency links are typically tracked for entire compilation units (files).
+
+One interesting point here is that Message Inlining creates delicate dependencies between object formats and method code. This is because if an object map changes, the compile-time search for the correct slot may be invalidated.
+
+### Support for Source-Level Debugging
+
+This section explains how extra state is maintained to construct the "virtual callstack" (the one the programmer probably has in mind) from the "physical callstack" (which excludes the inlined methods), and also how to resolve at which program counter to insert a breakpoint.
+
+## Performance Comparison
+
+There's a lot to unpack here, and to save some fuel for discussion, I'll just outline a few important points.
+
+* The authors acheive roughly a 2x speedup from the fastest Smalltalk-80 system, both on large standard C benchmarks (which are transliterated into Self and Smalltalk), and on a handful of small programs which are manually written in Self.
+* The authors propose a new metric for measuring performance of Smalltalk-like languages, *Mims* (millions of messages per second) and a notion of *efficiency* (messages per instruction).
+* Self is also compared to two versions of Smalltalk which include type declarations, and it acheives roughly the same performance.
+
+## Merits and Shortcomings of the Paper
+
+I won't do anything other than heap praise on the Self implementation described in the paper. There are a lot of great ideas, some of which advance prototype-based languages and some of which improve upon Smalltalk-like languages.
```

Comment:
Sorry if this sounds blunt, but this is a bit of a cop-out. You don't need to have a biting critique of the paper or anything, or even have a separate section for your opinions. But just integrating some of your perspective, to complement the direct summarization, would help quite a bit. You have already done some of this in other sections, so no need to say "I don't have any options"—just include some critical thinking where it's appropriate.